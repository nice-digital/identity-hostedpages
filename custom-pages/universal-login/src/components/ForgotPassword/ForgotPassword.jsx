import React, { Component } from "react";
import { NavLink, Link } from "react-router-dom";
import { Alert } from "@nice-digital/nds-alert";
import { Input } from "@nice-digital/nds-input";

import { AuthApi } from "../../services/AuthApi";
import { isDomainInUsername, validateRegisterFields } from "../../helpers";

import "./ForgotPassword.scss";

export class ForgotPassword extends Component {
  constructor(props) {
    super(props)
    this.auth = new AuthApi()
    this.state = {
      message:  (props.location && props.location.state && props.location.state?.message) || null,
      value: (props.location && props.location.state && props.location.state?.email) || null,
      errors: {
        email: false
      },
      serverSideError: null,
      loading: false,
      isAD: false
    }
  }

  forgotPassword = (event) => {
    if (event) event.preventDefault();
    if(this.state.value)
    {
       const simulatedEvent = { target: document.getElementById('email') };
       this.handleChange(simulatedEvent);
    }

    this.setState(function(state) {
      const tests = validateRegisterFields(this.state);
      const email = !state.email || tests.email();

      return {
        errors: {
          email: email
        }
      };
    }, () => {
      if (!this.state.errors.email) {
        this.doForgotPassword();
      } else {
        this.setState({ serverSideError: null });
      }
    });
  };

  doForgotPassword = () => {
    const serverErrorCallback = err => {
      console.error(err);

      this.setState({
        serverSideError: err.description || err.name,
        loading: false
      });
    };

    try {
      this.setState({ loading: true });

      this.auth.forgotPassword(
        this.state.email,
        serverErrorCallback,
        this.props.history
      );
    } catch (err) {
      console.error(err);

      this.setState({
        serverSideError: err.message || err.name,
        loading: false
      });
    }
  };

  handleChange = (event) => {
    let name = event.target.name,
      value = event.target.value;

    this.setState(function(state) {
      let isAD = (name === 'email') ? isDomainInUsername(value) : state.isAD;

      return {
        [name]: value,
        serverSideError: null,
        isAD
      };
    }, () => {
      if (this.state.errors.email) {
        this.validate();
      }
    });
  };

  clearError = (event) => {
    let eventTargetName = event.target.name;

    this.setState(function(state) {
      return {
        errors: { ...state.errors, [eventTargetName]: false },
        serverSideError: null
      };
    });
  };

  validate = () => {
    const tests = validateRegisterFields(this.state);
    const email =  this.state.email ? tests.email(this.state.email) : this.state.errors.email;

    this.setState({
      errors: {
        email: email,
      }
    });
  };

  render() {
    const { serverSideError, errors, loading, email, isAD, message } = this.state;

    return (
      <div>
        <h2 className="mt--0">Reset your password</h2>
        {message &&
        	<p class="alert alert--error">Our password policy has been updated. You will need to provide a password that is at least 14 characters in length and contains at least 3 of the following 4 types of characters: lower case letters (a-z), upper case letters (A-Z), numbers (i.e. 0-9) and special characters (e.g. !@#$%^&*). <br/><br/>Click the reset button below and we'll send you an email with a link to help you reset your password.</p>
        }
        {!this.state.value && (
					<p className="lead">
						Enter the email address you registered with in the box below and click the reset button. We'll send you an email with a link to help you reset your password.
					</p>
         )}
        <form className="">
          {serverSideError && (
            <Alert type="error">{serverSideError}</Alert>
          )}
          <Input
            data-qa-sel="forgotPassword-email"
            label="Email"
            id="email"
            name="email"
            unique="email"
            type="email"
            placeholder="eg: your.name@example.com..."
            onChange={this.handleChange}
            error={errors.email}
            errorMessage={`${
              !email
                ? 'This field is required'
                : 'Email address is in an invalid format'
            }`}
            onBlur={this.validate}
            // onFocus={this.clearError}
            aria-describedby="email-error"
            value={this.state.value}
          />
          {isAD ? (
            <Alert type="info">
              NICE staff should <NavLink data-qa-sel="Signin-link-login" to="/" activeclassname="activeRoute">sign in</NavLink> using the password you use to sign in to your work computer.
            </Alert>
          ) : (
          !loading ? (
            <button
              data-qa-sel="forgotPassword-button"
              className="btn btn--cta"
              onClick={this.forgotPassword}
            >
              Reset password
            </button>
          ) : (
            'Loading...'
          )
          )}
        </form>
        {!isAD &&
        <Link data-qa-sel="forgotPassword-link-to-login" to="/">
          Return to sign in
        </Link>
        }
      </div>

    )
  }
}
