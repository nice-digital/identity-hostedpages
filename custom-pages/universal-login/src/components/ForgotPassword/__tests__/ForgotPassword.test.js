import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import { ForgotPassword } from "../ForgotPassword";

jest.mock("../../../helpers/isLoginPage", () => true);
jest.mock("../../../services/AuthApi")

describe("ForgotPassword components", () => {
  // let el
  // let instance
  const auth = {
    forgotPassword: jest.fn()
  }
  //const functionSignature = "a function signature"

  // beforeEach(() => {
  //   el = shallow(<ForgotPassword />)
  //   instance = el.instance()
  //   instance.auth = auth
  //   instance.requestErrorCallback = functionSignature
  // })

  it("should render <ForgotPassword /> correctly", () => {
		const { container } = render(
			<MemoryRouter future={{ v7_relativeSplatPath: true, v7_startTransition: true }}>
				<ForgotPassword />
			</MemoryRouter>
		);
    expect(container).toMatchSnapshot();
  });

  it("should render correctly when there is an error", async () => {
		render(
			<MemoryRouter future={{ v7_relativeSplatPath: true, v7_startTransition: true }}>
				<ForgotPassword />
			</MemoryRouter>
		);
		const resetPasswordButton = await screen.findByRole("button", { name: "Reset password"});
		userEvent.click(resetPasswordButton);
		const errorAlert = await screen.findByText("This field is required");
  	expect(errorAlert).toBeInTheDocument();
  });
})
