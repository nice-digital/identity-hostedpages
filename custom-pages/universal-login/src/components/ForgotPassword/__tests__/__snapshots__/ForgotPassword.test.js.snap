// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForgotPassword components should render <ForgotPassword /> correctly 1`] = `
<div>
  <div>
    <h2
      class="mt--0"
    >
      Reset your password
    </h2>
    <p
      class="lead"
    >
      Enter the email address you registered with in the box below and click the reset button. We'll send you an email with a link to help you reset your password.
    </p>
    <form
      class=""
    >
      <div
        class="input"
        data-component="input"
      >
        <label
          class="input__label"
          for="email"
        >
          Email
        </label>
        <input
          aria-describedby="email-error"
          class="input__input"
          data-qa-sel="forgotPassword-email"
          id="email"
          name="email"
          placeholder="eg: your.name@example.com..."
          type="email"
          unique="email"
          value=""
        />
      </div>
      <button
        class="btn btn--cta"
        data-qa-sel="forgotPassword-button"
      >
        Reset password
      </button>
    </form>
    <a
      data-qa-sel="forgotPassword-link-to-login"
      href="/"
    >
      Return to sign in
    </a>
  </div>
</div>
`;
